backend/cfg/dataflow.ml

```diff
@@ -0,0 +1,218 @@
+open! Int_replace_polymorphic_compare
+
+module type Domain = sig
+  type t
+
+  val bot : t
+
+  val equal : t -> t -> bool
+
+  val lessequal : t -> t -> bool
+
+  val join : t -> t -> t
+
+  val to_string : t -> string
+end
+
+module type Transfer = sig
+  type domain
+
+  val basic : domain -> Cfg.basic Cfg.instruction -> domain
+
+  val terminator : domain -> Cfg.terminator Cfg.instruction -> domain
+
+  val exception_ : domain -> domain
+end
+
+module type S = sig
+  type domain
+
+  type value =
+    { before : domain;
+      after : domain option;
+      exception_ : domain option
+    }
+
+  val run :
+    Cfg.t ->
+    ?max_iteration:int ->
+    init:(Cfg.basic_block -> domain * bool) ->
+    unit ->
+    value Label.Tbl.t
+end
+
+module Forward (D : Domain) (T : Transfer with type domain = D.t) :
+  S with type domain = D.t = struct
+  type domain = D.t
+
+  type value =
+    { before : domain;
+      after : domain option;
+      exception_ : domain option
+    }
+
+  let transfer_block : value -> Cfg.basic_block -> value =
+   fun value block ->
+    let after =
+      Some
+        (T.terminator
+           (ListLabels.fold_left block.body ~init:value.before ~f:T.basic)
+           block.terminator)
+    in
+    let exception_ = Some (T.exception_ value.before) in
+    { value with after; exception_ }
+
+  let create :
+      Cfg.t ->
+      init:(Cfg.basic_block -> domain * bool) ->
+      value Label.Tbl.t * Label.Set.t ref =
+   fun cfg ~init ->
+    (* CR xclerc for xclerc: what should be the initial size? *)
+    let map = Label.Tbl.create 32 in
+    let set = ref Label.Set.empty in
+    Cfg.iter_blocks cfg ~f:(fun label block ->
+        let before, in_work_set = init block in
+        let after = None in
+        let exception_ = None in
+        Label.Tbl.replace map label { before; after; exception_ };
+        if in_work_set then set := Label.Set.add label !set);
+    map, set
+
+  let remove_and_return : Cfg.t -> Label.Set.t ref -> Cfg.basic_block =
+   fun cfg set ->
+    let label = Label.Set.choose !set in
+    set := Label.Set.remove label !set;
+    Cfg.get_block_exn cfg label
+
+  let different : domain option -> domain option -> bool =
+   fun left right -> not (Option.equal D.equal left right)
+
+  let run :
+      Cfg.t ->
+      ?max_iteration:int ->
+      init:(Cfg.basic_block -> domain * bool) ->
+      unit ->
+      value Label.Tbl.t =
+   fun cfg ?(max_iteration = max_int) ~init () ->
+    let res, work_set = create cfg ~init in
+    let iteration = ref 0 in
+    while (not (Label.Set.is_empty !work_set)) && !iteration < max_iteration do
+      incr iteration;
+      let block = remove_and_return cfg work_set in
```

Comment:
Have you considered storing  a pair of block label and abstract value in the workset, instead of just a label? I think it would simplify the code in L102-L119, and would mean there is only one abstract value to store per block, instead of three (in L49-51).